### PR TITLE
NPM Packages Table on Linux

### PR DESCRIPTION
**Pull request by Exbia**
Adding query support and table for NPM packages on Linux

### CODE REVIEW

**Comment by obelisk** [comment]
ok to test

File: osquery/tables/system/linux/npm_packages.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+#include "osquery/core/conversions.h"
+
+namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+const std::map<std::string, std::string> PackageTopLevelKeys{
```
Comment (obelisk) [inline osquery/tables/system/linux/npm_packages.cpp]: constants should start with a k, ie kPackageTopLevelKeys

File: osquery/tables/system/linux/npm_packages.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#include <boost/filesystem.hpp>
+
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+
+namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
```
Comment (obelisk) [inline osquery/tables/system/linux/npm_packages.cpp]: Please don't use ptree anymore and use the RapidJSON functions in conversions.

File: osquery/tables/system/linux/npm_packages.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    {"license", "license"}};
+
+QueryData genNPMPackages(QueryContext& context) {
+  QueryData results;
+
+  std::vector<std::string> paths;
+  resolveFilePattern("/usr/lib/node_modules/%/package.json", paths);
+
+  for (const auto& path : paths) {
+    pt::ptree tree;
```
Comment (obelisk) [inline osquery/tables/system/linux/npm_packages.cpp]: Please use JSON documents

File: osquery/tables/system/linux/npm_packages.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    {"version", "version"},
+    {"description", "description"},
+    {"author.name", "author"},
+    {"license", "license"}};
+
+QueryData genNPMPackages(QueryContext& context) {
+  QueryData results;
+
+  std::vector<std::string> paths;
+  resolveFilePattern("/usr/lib/node_modules/%/package.json", paths);
```
Comment (obelisk) [inline osquery/tables/system/linux/npm_packages.cpp]: If you have this passed in via `WHERE` clause then we can get this info from any directory on the system or fall back to global is one isn't provided.

**Comment by Exbia** [comment]
I've removed the use of ptree for reading the package.json file, and re-implemented it with JSON Documents. 

I've also added a 'directory' column which can be used in WHERE filters.

With these changes, we can now query node modules (globally installed or within a directory).

```
select * from npm_packages;
select * from npm_packages where directory = "/home/andrew/test_project/";
```

Further work will need to be done for other operating systems.

File: specs/linux/npm_packages.table [hunk 5]
```diff
@@ -0,0 +1,12 @@
+table_name("npm_packages")
+description("Lists all npm packages globally installed in a system.")
```
Comment (obelisk) [inline specs/linux/npm_packages.table]: Or installed in a directory

File: osquery/tables/system/linux/npm_packages.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    auto doc = JSON::newObject();
+    if (!doc.fromString(json) || !doc.doc().IsObject()) {
+      LOG(WARNING) << "Could not parse JSON from: " << package_path;
+      continue;
+    }
+
+    Row r;
+    for (const auto& key_it : kPackageKeys) {
+      std::string key{key_it};
```
Comment (obelisk) [inline osquery/tables/system/linux/npm_packages.cpp]: Do you need this copy here?

File: specs/linux/npm_packages.table [hunk 7]
```diff
@@ -0,0 +1,12 @@
+table_name("npm_packages")
+description("Lists all npm packages in a directory or globally installed in a system.")
+schema([
+    Column("name", TEXT, "Package display name"),
+    Column("version", TEXT, "Package supplied version"),
+    Column("description", TEXT, "Package supplied description"),
+    Column("author", TEXT, "Package author name"),
+    Column("license", TEXT, "License for package"),
+    Column("path", TEXT, "Path at to this module package.json"),
```
Comment (obelisk) [inline specs/linux/npm_packages.table]: ~at~ module's

**Comment by osqueryer** [comment]
:-1: The commit e866f65e9ee23932630df51d49b20a66ef5f010f (Job results: [7011](https://jenkins.osquery.io/job/osqueryPullRequestLinux/7011/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit e866f65e9ee23932630df51d49b20a66ef5f010f (Job results: [7012](https://jenkins.osquery.io/job/osqueryPullRequestLinux/7012/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 374a3ee899cbe52f59bd1a6e49d7c0b391930a04 (Job results: [3052](https://jenkins.osquery.io/job/osqueryPullRequestOSX/3052/)) failed one or more tests (**macOS/OS X**).

**Comment by theopolis** [comment]
> obelisk added the do not merge  label 6 hours ago

Curious as to why?

**Comment by obelisk** [comment]
@theopolis He is waiting to get the diff approved from his company

File: osquery/tables/system/linux/npm_packages.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    if (!doc.fromString(json) || !doc.doc().IsObject()) {
+      LOG(WARNING) << "Could not parse JSON from: " << package_path;
+      continue;
+    }
+
+    Row r;
+    for (const auto& key : kPackageKeys) {
+      if (doc.doc().HasMember(key)) {
+        const auto& value = doc.doc()[key];
+        r[key] = (value.IsString()) ? value.GetString() : "";
```
[the remark this answers is not in the document]
Comment (Exbia) [reply osquery/tables/system/linux/npm_packages.cpp]: Each key here should be a string. I think you're right, if a key is not the type expected then maybe we discard the credibility of the file and skip it with a log message.

**Comment by osqueryer** [comment]
:-1: The commit 1c74a5e545c297eb42f245e66fc88c5f4ecd0571 (Job results: [3063](https://jenkins.osquery.io/job/osqueryPullRequestOSX/3063/)) failed one or more tests (**macOS/OS X**).

File: osquery/tables/system/linux/npm_packages.cpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        const auto& value = doc.doc()[key];
+        // npm has a schema for package.json, but it is loosely enforced. Some
+        // keys may be missing, so populate the columns we can
+        r[key] = (value.IsString()) ? value.GetString() : "";
+      }
+    }
+
+    // Manually get nested key (Author name)
+    if (doc.doc().HasMember("author")) {
+      const auto& author = doc.doc()["author"]["name"];
```
Comment (fmanco) [inline osquery/tables/system/linux/npm_packages.cpp]: Should you be checking the key "name" exists before accessing it?

File: osquery/tables/system/linux/npm_packages.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        if (license.HasMember("url")) {
+          const auto& license_url = doc.doc()["license"]["url"];
+          if (license_url.IsString()) {
+            // Fallback to displaying deprecated licence url
+            r["license"] = license_url.GetString();
+          }
+        }
+      }
+
+      const auto& author = doc.doc()["author"]["name"];
```
Comment (fmanco) [inline osquery/tables/system/linux/npm_packages.cpp]: Is this supposed to be repeated here?

File: osquery/tables/system/linux/npm_packages.cpp [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    // dictionary with url and type
+    if (doc.doc().HasMember("license")) {
+      const auto& license = doc.doc()["license"];
+      if (license.IsString()) {
+        // Current license schema is just a top level string
+        r["license"] = license.GetString();
+      } else {
+        // If its not a string, is it a dict with 'url' ?
+        if (license.HasMember("url")) {
+          const auto& license_url = doc.doc()["license"]["url"];
```
Comment (fmanco) [inline osquery/tables/system/linux/npm_packages.cpp]: I didn't look up the interface but can you do `license["url"]`?

File: osquery/tables/system/linux/npm_packages.cpp (hunk 9 above)
Comment (Exbia) [reply osquery/tables/system/linux/npm_packages.cpp]: Yes, I've fixed it now

File: osquery/tables/system/linux/npm_packages.cpp (hunk 10 above)
Comment (Exbia) [reply osquery/tables/system/linux/npm_packages.cpp]: No it was not, fixed

File: osquery/tables/system/linux/npm_packages.cpp (hunk 11 above)
Comment (Exbia) [reply osquery/tables/system/linux/npm_packages.cpp]: Yup, much more readable too, fixed

**Pull request merged**